pytket_qir/converter.py

```diff
@@ -0,0 +1,357 @@
+from collections import OrderedDict
+from dataclasses import dataclass
+import os
+from typing import Optional, cast, List, Union
+
+from pytket import Circuit  # type: ignore
+from pytket.circuit import CircBox  # type: ignore
+from pytket.circuit.logic_exp import BitNot, if_bit  # type: ignore
+from pytket.wasm import WasmFileHandler  # type: ignore
+
+from pyqir.parser import (  # type: ignore
+    QirBlock,
+    QirBrTerminator,
+    QirCondBrTerminator,
+    QirLocalOperand,
+    QirModule,
+    QirRetTerminator,
+)
+from pyqir.generator import ir_to_bitcode, types  # type: ignore
+
+from pytket_qir.gatesets.base import CustomGateSet
+from pytket_qir.parser import QirParser
+from pytket_qir.utils import ConverterException
+
+
+@dataclass
+class Block:
+    name: str
+    succs: List
+    preds: List
+    composition: List
+    visited: bool = False
+
+
+def topological_sort(digraph: dict) -> dict:
+    """Topologically sort a digraph, a dict of <nodes, list of successors>."""
+    # Mapping nodes to indegrees.
+    indegrees = {node: 0 for node in digraph}
+    for node in digraph:
+        for succ in digraph[node]:
+            indegrees[succ] += 1
+
+    nodes_with_empty_preds = []
+    for node in digraph:
+        if indegrees[node] == 0:
+            nodes_with_empty_preds.append(node)
+
+    topological_ordering = {}
+    while len(nodes_with_empty_preds) > 0:
+        node = nodes_with_empty_preds.pop()
+        topological_ordering[node] = digraph[node]
+        for succ in digraph[node]:
+            indegrees[succ] -= 1
+            if indegrees[succ] == 0:
+                nodes_with_empty_preds.append(succ)
+
+    if len(topological_ordering) == len(digraph):
+        return topological_ordering
+    else:
+        raise ConverterException(
+            "A cycle exists in the graph, there is no topological order to be found."
+        )
+
+
+class QirConverter:
+    """
+    A class to build and convert the QIR control-flow graph to a pytket circuit.
+    """
+
+    def __init__(
+        self,
+        file_path: str,
+        optimisation_level: int = 0,
+        gateset: Optional[CustomGateSet] = None,
+        wasm_handler: Optional[WasmFileHandler] = None,
+        wasm_int_type: types.Int = types.Int(32),
+        qir_int_type: types.Int = types.Int(64),
+    ) -> None:
+        self.module = QirModule(file_path)
+        self.module_function = self.module.functions[0]
+        self.parser = QirParser(
+            qir_module=self.module,
+            gateset=gateset,
+            wasm_handler=wasm_handler,
+            wasm_int_type=wasm_int_type,
+            qir_int_type=qir_int_type,
+        )
+        self.cfg = OrderedDict()
+        self.rewritten_cfg = self.cfg
+        if optimisation_level == 1:
+            self.collapse_blocks()
+        self.circuit = self.cfg_to_circuit()
+
+    @property
+    def successors(self):
+        return self._successors
+
+    @successors.setter
+    def successors(self, value):
```

Comment:
Can you maybe add a comment for what this is used?

Reply:
Done.